app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt

```diff
@@ -0,0 +1,145 @@
+        "www.test.com",
+        "www.test.com",
+        "CNN",
+        "Show me how to test.",
+    )
+
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    var coroutinesTestRule = CoroutinesTestRule()
+
+    @get:Rule
+    var instantTaskExecutorRule = InstantTaskExecutorRule()
+
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+
+        // FIXME: can we put every in set up before test start?
+//        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+//        viewModel = NewsViewModel(mockRepository)
+    }
+
+    @Test
+    fun `When calling print message actually prints out the message`() {
+
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        assertEquals("This is in test", viewModel.getMessage("This is in test"))
+
+        verify(exactly = 1) { mockRepository.isDataUsage() }
+//        verify(exactly = 1) { viewModel.print("123") }
+    }
+
+    @Test
+    fun `print message`() {
+
+        every { mockRepository.isDataUsage() } answers {
+            flow { }
+//            flow { emit(true) }
+        }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.print("1234")
+
+        verify(exactly = 1) { viewModel.print("1234") }
+//        verify(exactly = 1) { println("1234") }
+    }
+
+    @Test
+    fun `fetch articles`() {
+
+        val mockObserver = mockk<Observer<ArticleState>>()
+
+        val expectedResult = ArticleState.Ready(listOf(dummyArticle))
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        every { mockRepository.getArticles() } returns flow { emit(expectedResult) }
+
+        var onChangedInvoked = 0
+
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.headLineNewsLiveData.observeForever(mockObserver)
+
+        every { mockObserver.onChanged(expectedResult) } answers {
+            onChangedInvoked++
+        }
+
+        viewModel.fetchArticle()
+
+        verify(exactly = 1) { mockRepository.getArticles() }
+
+//        assertEquals(1, onChangedInvoked)
+        verify { mockObserver.onChanged(expectedResult) }
+    }
+
+    @Test
+    fun `search articles`() {
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.searchArticles("test")
+
+        coVerify(exactly = 1) { mockRepository.searchArticles(any()) }
+    }
+
+    @Test
+    fun `toggle data usage`() {
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.toggleDataUsage()
+
+        coVerify(exactly = 1) { mockRepository.toggleDataUsage() }
+    }
+}
+
+/**
+ * Question:
+ * 1. how do I test the function that returns Unit (nothing)
+ * 2. what is the difference between black-box and white-box testing
```

Comment:
Black box testing involves testing that has _no understanding_ of the inner workings and software components of the system being tested.  Appium is a good example of a black box test framework for mobile platforms.  You write your test code to "simulate" a user interaction and it runs those commands on the app. For example, tap a button, enter email, tap login button, enter password, etc.  These tests are typically written by testers or QA.

White box testing means you have a good understanding of the software and are testing individual pieces of the inner software.  Integration testing and unit testing are good examples here.  These tests are written by the software developers themselves.